File: src/main/java/com/autosale/shop/repository/ProductRepository.java
package com.autosale.shop.repository;

import com.autosale.shop.model.PaginationRequest;
import com.autosale.shop.model.Product;

import java.util.List;
import java.util.Optional;

public interface ProductRepository {

    Optional<Integer> save(Product product);

    Optional<Product> findById(int id);

    void update(Product product);

    int deleteById(int id);

    int deleteByIdInArray(List<Integer> id);

    List<Product> findAllByUserId(int userId);

    List<Product> findAllByStatus(PaginationRequest pageRequest, String status);

    int countAllByStatus(PaginationRequest pageRequest, String status);

    void saveAllIgnoreExistence(List<Product> product);
}

File: src/main/java/com/autosale/shop/repository/ProductRepository.kt
package com.autosale.shop.repository

import com.autosale.shop.model.PaginationRequest
import com.autosale.shop.model.Product
import com.autosale.shop.model.ProductStatus

interface ProductRepository {
    fun save(product: Product): Int
    fun findById(id: Int): Product?
    fun update(product: Product)
    fun deleteById(id: Int): Int
    fun deleteByIdInArray(ids: List<Int>): Int
    fun findAllByUserId(userId: Int): List<Product>
    fun findAllByStatus(pageRequest: PaginationRequest, status: String?): List<Product>
    fun countAllByStatus(pageRequest: PaginationRequest, status: String?): Int
    fun saveAllIgnoreExistence(products: List<Product>)
    fun setStatus(productId: Int, status: ProductStatus)
    fun countAllActive(): Int
}
